In [2]:
import pandas as pd
import numpy as np
import csv

In [3]:
data = pd.read_csv('광진구 도로폭.csv',encoding='CP949')

In [4]:
data.columns

Index(['시군구', '위계', '광역구분', '종속구분', '도로명', '영문도로명', '고시일자', '시점', '종점', '도로폭',
       '도로고시길이', '도로물리길이', '기초간격', '부여사유', '부여방식', '부여일자', '부여사유설명',
       '도로구간일련번호'],
      dtype='object')

In [5]:
wide = data[['도로명','도로폭']]

In [6]:
wide

도로명  도로폭
0        천호대로   52
1        강변북로   20
2        강변역로   39
3        광나루로   38
4         광장로   15
..        ...  ...
538  천호대로140길   10
539  천호대로141길    7
540  천호대로143길    9
541  천호대로145길    6
542   한강북자전거길    2

[543 rows x 2 columns]

In [7]:
wide_under_6 = wide[wide['도로폭'] <=6]

In [8]:
wide_under_6.count()

도로명    245
도로폭    245
dtype: int64

In [9]:
wide_under_6.reset_index()

index       도로명  도로폭
0       21    강변역로2길    5
1       23   광나루로12길    6
2       24   광나루로13길    6
3       25   광나루로14길    6
4       27   광나루로16길    5
..     ...       ...  ...
240    531  천호대로131길    6
241    534  천호대로135길    5
242    537  천호대로138길    5
243    541  천호대로145길    6
244    542   한강북자전거길    2

[245 rows x 3 columns]

In [26]:
wide_under_6.to_csv('wide_under_6.csv',encoding = 'CP949')
wide.to_csv('wide.csv',encoding = 'CP949')

In [1]:
from geopy.geocoders import Nominatim

In [34]:
def geocoding(address):
    while True:
        geolocoder = Nominatim(user_agent = 'South Korea', timeout=None)
        geo = geolocoder.geocode(address)
        if geo == None:
            print(address)
            a = "1"
            return a
            break
        crd = [str(geo.latitude), str(geo.longitude)]#float로 하면 메모리 많이 잡아먹으니까 str으로 저장하고 사용할 때 float로 변경
        return crd
        break

In [35]:
geo_x_y = []
for i in wide['도로명']:
    a = geocoding(i)
    if a == "1":#지오코드로 위도경도가 없다면
        geo_x_y.append(geo_x_y[-1])#바로 이전에 추가한 위도경도를 추가한다.
    else:
        geo_x_y.append(a)

광나루로12길
광장로1나길
광장로7길
광장로9길
구의로10길
구의로12길
구의로14길
구의로1길
구의로8길
군자로9길
긴고랑로1길
긴고랑로27길
긴고랑로46가길
긴고랑로46나길
능동로11길
능동로1길
능동로3나길
동일로10가길
동일로10나길
뚝섬로28길
뚝섬로30가길
뚝섬로32가길
뚝섬로67가길
면목로15길
아차산로39가길
아차산로41가길
아차산로41길
아차산로59길
아차산로63길
영화사로20길
자양로19가길
자양로25길
자양번영로10길
중랑천동자전거길
천호대로131길
한강북자전거길


## 지오코드를 활용해서 광진구 내의 도로명 주소들의 경도위도 좌표를 구했다.
## 하지만 도로명 주소로 경도위도가 나오지 않는 경우가 있었다. 이에 대한 해결책으로
## ㄱㄴㄷ순으로 정렬된 도로명 주소의 특징을 활용하여 경도위도가 없을 경우 바로 앞의 경도 위도 값을 대입하였다.
## 일부 예외는 수작업으로 수정. - > 카카오 api

위도경도가 나오지 않은 도로명 주소
광나루로12길
광장로1나길
광장로7길
광장로9길
구의로10길
구의로12길
구의로14길
구의로1길
구의로8길
군자로9길
긴고랑로1길
긴고랑로27길
긴고랑로46가길
긴고랑로46나길
능동로11길
능동로1길
능동로3나길
동일로10가길
동일로10나길
뚝섬로28길
뚝섬로30가길
뚝섬로32가길
뚝섬로67가길
면목로15길
아차산로39가길
아차산로41가길
아차산로41길
아차산로59길
아차산로63길
영화사로20길
자양로19가길
자양로25길
자양번영로10길
중랑천동자전거길
천호대로131길
한강북자전거길

In [36]:
geo_x_y

[['37.5466502', '127.1732408'],
 ['37.3195968', '127.6284452'],
 ['37.5362962', '127.0941979'],
 ['37.548316', '127.0665212'],
 ['35.1626277', '128.9807519'],
 ['37.53552', '127.09211'],
 ['37.54152', '127.08992'],
 ['37.5446913', '127.1134486'],
 ['37.3640899', '126.7686518'],
 ['37.56369', '127.0782554'],
 ['37.5657003', '127.0841887'],
 ['37.7074535', '127.0562534'],
 ['37.6242653', '127.0734496'],
 ['37.5427804', '127.021923'],
 ['37.5925264', '127.0864975'],
 ['37.5407603', '127.0679553'],
 ['37.5525116', '127.0995989'],
 ['37.6137974', '127.1014054'],
 ['37.5439809', '127.0972081'],
 ['37.5411813', '127.0836592'],
 ['37.5356431', '127.0764146'],
 ['37.5372643', '127.0911827'],
 ['37.5352829', '127.09379'],
 ['37.5352829', '127.09379'],
 ['37.5490815', '127.0685421'],
 ['37.5475771', '127.0687855'],
 ['37.5498847', '127.069671'],
 ['37.5470944', '127.0693319'],
 ['37.5487957', '127.0727108'],
 ['37.5467058', '127.0721021'],
 ['37.5483513', '127.0732359'],
 ['37.5452096', '127.0749

In [37]:
print(len(wide['도로명']))
print(len(geo_x_y))

543
543


In [39]:
wide = wide.copy()
wide['위도경도'] = geo_x_y

In [40]:
wide.head()

도로명  도로폭                       위도경도
0  천호대로   52  [37.5466502, 127.1732408]
1  강변북로   20  [37.3195968, 127.6284452]
2  강변역로   39  [37.5362962, 127.0941979]
3  광나루로   38   [37.548316, 127.0665212]
4   광장로   15  [35.1626277, 128.9807519]

In [41]:
wide_under_6 = wide[wide['도로폭'] <=6]

In [43]:
wide_under_6.head()

도로명  도로폭                       위도경도
21   강변역로2길    5  [37.5372643, 127.0911827]
23  광나루로12길    6    [37.5352829, 127.09379]
24  광나루로13길    6  [37.5490815, 127.0685421]
25  광나루로14길    6  [37.5475771, 127.0687855]
27  광나루로16길    5  [37.5470944, 127.0693319]

## 지오코딩으로 구하지 못한 것들은 카카오api를 활용하자.

In [44]:
import requests, json

In [45]:
def get_location(address):
  url = 'https://dapi.kakao.com/v2/local/search/address.json?query=' + address
  # 'KaKaoAK '는 그대로 두시고 개인키만 지우고 입력해 주세요.
  # ex) KakaoAK 6af8d4826f0e56c54bc794fa8a294
  headers = {"Authorization": "KakaoAK 1f43da856082215c9e7758b0e1c075a4"}
  api_json = json.loads(str(requests.get(url,headers=headers).text))
  address = api_json['documents'][0]['address']
  crd = [str(address['y']), str(address['x'])]
  address_name = address['address_name']

  return crd


In [46]:
geo_x_y = []
for i in wide['도로명']:
    a = geocoding(i)
    if a == "1":#지오코드로 위도경도가 없다면
        geo_x_y.append(geo_x_y[-1])#바로 이전에 추가한 위도경도를 추가한다.
    else:
        geo_x_y.append(a)

광나루로12길
광장로1나길
광장로7길
광장로9길
구의로10길
구의로12길
구의로14길
구의로1길
구의로8길
군자로9길
긴고랑로1길
긴고랑로27길
긴고랑로46가길
긴고랑로46나길
능동로11길
능동로1길
능동로3나길
동일로10가길
동일로10나길
뚝섬로28길
뚝섬로30가길
뚝섬로32가길
뚝섬로67가길
면목로15길
아차산로39가길
아차산로41가길
아차산로41길
아차산로59길
아차산로63길
영화사로20길
자양로19가길
자양로25길
자양번영로10길
중랑천동자전거길
천호대로131길
한강북자전거길


### 카카오나 지오나 똑같다

==> 다시 수작업

데이터가 더 필요하다면 
119 긴급신고 접수현황 https://www.bigdata-119.kr/goods/goodsInfo?goods_id=202212000063
화재출동 현황 https://www.bigdata-119.kr/goods/goodsInfo?goods_id=202212000070

In [48]:
wide_under_6.to_csv('wide_under_6_xy.csv',encoding = 'CP949')
wide.to_csv('wide.csv_xy.csv',encoding = 'CP949')

In [49]:
data = pd.read_csv('광진구 도로폭 위도경도수정.csv',encoding='CP949')

# 지도에 시각화

In [52]:
import folium

In [55]:
data.head()

Unnamed: 0   도로명  도로폭                           위도경도
0           0  천호대로   52  ['37.5466502', '127.1732408']
1           1  강변북로   20  ['37.3195968', '127.6284452']
2           2  강변역로   39  ['37.5362962', '127.0941979']
3           3  광나루로   38   ['37.548316', '127.0665212']
4           4   광장로   15  ['35.1626277', '128.9807519']

In [58]:
for i in data[:3]:
    print(i)

Unnamed: 0
도로명
도로폭
위도경도


In [54]:
m = folium.Map(location = [map(float,data['위도경도'][0])], zoom_start=14)
m

ValueError: Expected two (lat, lon) values for location, instead got: [<map object at 0x000001F04D8BD460>].